docs: added MADR for MeshRateLimit (#5333)
kumahq/kuma: docs/madr/decisions/011-rate-limit-policy.md

# Rate limit policy
 
- Status: accepted
 
Technical Story: https://github.com/kumahq/kuma/issues/4740
 
## Context and Problem Statement
 
We want to create a [new policy matching compliant](https://github.com/kumahq/kuma/blob/22c157d4adac7f518b1b49939c7e9ea4d2a1876c/docs/madr/decisions/005-policy-matching.md)
resource for managing rate limiting.
 
Rate Limiting in Kuma is implemented with Envoy's [rate limiting
support](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter).
There is an issue with the configuration of TCP rate limiting because [current filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/local_rate_limit_filter#config-network-filters-local-rate-limit) works on listener level which can only

> allows limiting connections on the specific listener but without checking the origin of the request.

Reviewer: ```suggestion
allow limiting connections on the specific listener but without checking the origin of the request.
```

     }
   }
   ]
 },
 "route": {
   "cluster": "localhost:5000",
   "timeout": "0s"
 },
 "typed_per_filter_config": {
   "envoy.filters.http.local_ratelimit": {
   "@type": "type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit",
   ...
   }
 }
}
 
```
 
### Specification
 
The differences between TCP and HTTP rate limiting in `targetRef` and `from` requires different validation of objects.
Idea is to create one policy but with `oneof` and use different validation methods for them.
 
#### **HTTP Rate Limit**
 
Rate limiting can be configured on both HTTP connection managers and routes. Thanks to the header `x-kuma-tags`, which is propagated and set by Envoy, we can recognize the origin of the HTTP request.
 
#### Top level
 
Top-level targetRef can have all available kinds:
```yaml
targetRef:
 kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
 name: ...
```
Rate limiting is an inbound policy so only `from` should be configured.
 
#### From level
 
```yaml
from:
 - targetRef:
     kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
     name: ...
```
 

> Matching on MeshGatewayRoute and MeshHTTPRoute does not make sense (there is any route that a request originates from).

Reviewer: ```suggestion
Matching on MeshGatewayRoute and MeshHTTPRoute does not make sense (there isn't a route that a request originates from).
```

Reviewer: @michaelbeaumont for a parallel discussion out of scope of this MADR. How do we envision this in the future with route? This is a use case for route to be defined on the producer side (I want to define a rate-limit for a subset of my api).

Reviewer 2: wouldn't you use MeshHTTPRoute in top-level targetRef to do so?

 
#### Configuration
 
```yaml
 default:
   http:
     requests: 5
     interval: 10s
     onRateLimit:
       status: 423
       headers:
         - key: "x-kuma-rate-limited"
           value: "true"
           append: true
```
 
#### **TCP Connection Rate Limit**
 
Rate limiting on TCP connection is more complicated because there is no easy way to match only for specific clients.
Based on alternatives we decided to make it only available for all the clients.
 
#### Top level
 
```yaml
targetRef:
 kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
 name: ...
```
Matching on MeshGatewayRoute and MeshHTTPRoute does not make sense because we are configuring TCP listeners.
Rate limiting is an inbound policy so only `from` should be configured.
 
#### From level
 
```yaml
from:
 - targetRef:
     kind: Mesh
     name: ...
```
 

> There is no easy way to match requesting services with the specific filter chain that's why we decided to support only configuring TCP rate limiting from all services.

Reviewer: Is this the recurring issue (#1256) of missing propagation to further filter the service id extracted at the TLS level?

Reviewer: Actually seems like #1256 only applies to L7. @jakubdyszkiewicz there's no way at L4 to do this?

Author: At L7 it is easier but at L4 I couldn't find a solution.

Reviewer 2: I also don't see a reasonable way to do this on L4.

 
#### Configuration
 
```yaml
 default:
   tcp:
     connections: 100
     interval: 10s
```
 
#### **Result**
 
```yaml
type: MeshRateLimit 
mesh: example 
name: example-rate-limit
spec:
 targetRef:
   kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute # or Mesh|MeshSubset|MeshService|MeshServiceSubset for TCP configuration
   name: example-rate-limit
 from:
   - targetRef:
       kind: Mesh|MeshSubset|MeshService|MeshServiceSubset # or Mesh for TCP configuration
       name: backend
       mesh: example

>      default: # you can only define one: http or tcp

Author: It could be possible, I think we might do it and only allow both if there is Mesh

       http:
         requests: 5
         interval: 10s
         onRateLimit:
           status: 423
           headers:
             - key: "x-kuma-rate-limited"
               value: "true"
               append: true
       tcp:
         connections: 100
         interval: 10s
```
 
### Considered Options
 
#### TCP rate limit for a specific origin
 
Configuration of connection limits is done on listeners. In case of matching from which service a request arrived, we
need to get some information about the request. There are two options:
 
* use SourceIP filter matching
* use SNI filter matching
 
##### Use SourceIP matching
 
In this case, we need to configure a listener filter matcher with a list of all possible IPs from which requests can arrive. In the case of one service, it can be many IPs. Matching on many ips might not be the most efficient way and each change of IP requires listener reload.
 
##### Use SNI filter matching
 

> It requires a change of SNI that is set on cluster `service_name{mesh=name}` to something different `service_name{mesh=name}{origin=requesting_service}` to later match it in the filter chain. That is not the best idea and can cause issues in multi-zone setups. Apart from that, it requires mTLS to be enabled.

Reviewer: Having feature being unavailable because of mTLS not being on is not new and I think it's acceptable.
So the only way to do per connection origin is by using SNI? No way to get info from the SAN?

Author: Unfortunately, I haven't seen any way to do this with SAN. Filter matcher can only get it by source/destination/protocol/SNI. So there is no way to retrieve SAN.

 destinations:
   - match:
       kuma.io/service: "demo-app_kuma-demo_svc_5000"
 conf:
   http:
     requests: 5
     interval: 10s
     onRateLimit:
       status: 423
       headers:
         - key: "x-Kuma-rate-limited"
           value: "true"
           append: true
```
We are using [Envoy's local rate limit](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/local_rate_limit_filter) which supports only HTTP configuration.
 
The configuration translates to the Envoy configuration on the route and the listener filter:
 
```yaml
{
 "match": {
   "prefix": "/",
   "headers": [
   {
     "name": "x-kuma-tags",
     "safe_regex_match": {
     "google_re2": {},
     "regex": ".*&kuma.io/service=[^&]*redis_kuma-demo_svc_6379[,&].*"
     }
   }
   ]
 },
 "route": {
   "cluster": "localhost:5000",
   "timeout": "0s"
 },
 "typed_per_filter_config": {
   "envoy.filters.http.local_ratelimit": {
   "@type": "type.googleapis.com/envoy.extensions.filters.http.local_ratelimit.v3.LocalRateLimit",
   ...
   }
 }
}
 
```
 
### Specification
 
The differences between TCP and HTTP rate limiting in `targetRef` and `from` requires different validation of objects.

> Idea is to create one policy but and allow configuring TCP and HTTP rate limiting at once. The only limitation is that TCP rate limit can be configured only when `from` section has `Mesh` and `targetRef` can have following values `Mesh|MeshSubset|MeshService|MeshServiceSubset`.

Reviewer: ```suggestion
We create one policy which can configure both TCP and HTTP rate limiting. Because, we can't identify incoming traffic when using TCP users can only set the TCP section when using the `Mesh` type of targetRef in the `from` section. For http these types of targetRef can be used: `Mesh|MeshSubset|MeshService|MeshServiceSubset`.
```

> Idea is to create one policy but and allow configuring TCP and HTTP rate limiting at once. The only limitation is that TCP rate limit can be configured only when `from` section has `Mesh` and `targetRef` can have following values `Mesh|MeshSubset|MeshService|MeshServiceSubset`.

Author: I think this might be tricky, because I can see 2 cases(and symmetric for TCP):
* user wants to configure the service's http rate limit but the service isn't HTTP - we use TCP config
* user wants to configure the service's http rate limit and the service is HTTP - we use HTTP config

At the validation, we cannot check if the service is HTTP or TCP, just when we generate configuration.

> Idea is to create one policy but and allow configuring TCP and HTTP rate limiting at once. The only limitation is that TCP rate limit can be configured only when `from` section has `Mesh` and `targetRef` can have following values `Mesh|MeshSubset|MeshService|MeshServiceSubset`.

Reviewer: Is there any reason to not do both TCP and HTTP at the same time? Because if we allow this this doesn't become a problem anymore

> Idea is to create one policy but and allow configuring TCP and HTTP rate limiting at once. The only limitation is that TCP rate limit can be configured only when `from` section has `Mesh` and `targetRef` can have following values `Mesh|MeshSubset|MeshService|MeshServiceSubset`.

Author: I don't see any obvious reason to not do it, maybe more configuration.

> Idea is to create one policy but and allow configuring TCP and HTTP rate limiting at once. The only limitation is that TCP rate limit can be configured only when `from` section has `Mesh` and `targetRef` can have following values `Mesh|MeshSubset|MeshService|MeshServiceSubset`.

Reviewer 2: I think I'm ok with both

 
#### **HTTP Rate Limit**
 
Rate limiting can be configured on both HTTP connection managers and routes. Thanks to the header `x-kuma-tags`, which is propagated and set by Envoy, we can recognize the origin of the HTTP request.
 
#### Top level
 
Top-level targetRef can have all available kinds:
```yaml
targetRef:
 kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
 name: ...
```
Rate limiting is an inbound policy so only `from` should be configured.
 
#### From level
 
```yaml
from:
 - targetRef:
     kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
     name: ...
```
 
Matching on MeshGatewayRoute and MeshHTTPRoute does not make sense (there is any route that a request originates from).
 
#### Configuration
 
```yaml
default:
  local:
    http:
      requests: 5
      interval: 10s
      onRateLimit:
        status: 423
        headers:
          - key: "x-kuma-rate-limited"
            value: "true"

>             append: true

Reviewer: Is there a default for `onRateLimit` ? If yes could we have `x-kuma-rate-limited` but set the value to the policy name?

Reviewer 2: I don't know...
1) We can leak policy names to the outside world
2) What do we put if we merge config from multiple policies?

Author: We do not create a default configuration for `onRateLimit`, if status/headers are not defined, no headers are added, and use default status set by Envoy: `429`

Reviewer: Ok let's stick with this defualt behaviour

```
 
#### **Result**
 
```yaml
type: MeshRateLimit 
mesh: example 
name: example-rate-limit
spec:
 targetRef:
   kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute # or Mesh|MeshSubset|MeshService|MeshServiceSubset for TCP configuration
   name: example-rate-limit
 from:
   - targetRef:
       kind: Mesh|MeshSubset|MeshService|MeshServiceSubset # or Mesh for TCP configuration
       name: backend
       mesh: example
     default: 
       local:
         http:
           enabled: true
           requests: 5
           interval: 10s
           onRateLimit:
             status: 423
             headers:
               - key: "x-kuma-rate-limited"
                 value: "true"
                 append: true
         tcp:
           enabled: true
           connections: 100
           interval: 10s
```
 
### Considered Options
 
#### TCP rate limit for a specific origin
 
Configuration of connection limits is done on listeners. In case of matching from which service a request arrived, we
need to get some information about the request. There are two options:
 
* use SourceIP filter matching
* use SNI filter matching
 
##### Use SourceIP matching
 

> In this case, we need to configure a listener filter matcher with a list of all possible IPs from which requests can arrive. In the case of one service, it can be many IPs. Matching on many ips might not be the most efficient way and each change of IP requires listener reload.

Reviewer 2: and this becomes impossible with multizone, because of ZoneIngress/ZoneEgress. Unless we would use PROXY protocol which adds complexity.
